tough/src/editor/signed.rs

```diff
@@ -494,3 +523,166 @@ impl SignedRepository {
         Ok(())
     }
 }
+
+impl TargetsWalker for SignedDelegatedTargets {
+    fn targets(&self) -> HashMap<String, &Target> {
+        // There are multiple `Targets` roles here that may or may not be related,
+        // so find all of the `Target`s related to each role and combine them.
+        let mut targets_map = HashMap::new();
+        for targets in &self.roles {
+            targets_map.extend(targets.signed.signed.targets_map());
+        }
+        targets_map
+    }
+
+    fn consistent_snapshot(&self) -> bool {
+        self.consistent_snapshot
+    }
+}
+
+/// `TargetsWalker` is used to unify the logic related to copying and linking targets.
+/// `TargetsWalker`'s default implementation of `walk_targets()` and `target_path()` use
+/// the trait's `targets()` and `consistent_snapshot()` methods to get a map of targets and
+/// also determine if a file prefix needs to be used.
+trait TargetsWalker {
+    /// Returns a map of all targets this manager is responsible for
+    fn targets(&self) -> HashMap<String, &Target>;
+    /// Determines whether or not consistent snapshot filenames should be used
+    fn consistent_snapshot(&self) -> bool;
+
+    /// Walks a given directory and calls the provided function with every file found.
+    /// The function is given the file path, the output directory where the user expects
+    /// it to go, and optionally a desired filename.
+    fn walk_targets<F>(
+        &self,
+        indir: &Path,
+        outdir: &Path,
+        f: F,
+        replace_behavior: PathExists,
```

Comment:
Maybe a little strange to accept replace_behavior only to have it passed to the function that was also passed. The repace_behavior could be captured or curried into `f`.

Not a blocker, just an observation.